apps/extension/src/ui/hooks/useResolveEnsName.ts

```diff
@@ -0,0 +1,51 @@
+import { isEthereumAddress } from "@polkadot/util-crypto"
+import { isPotentialEns } from "@talismn/on-chain-id"
+import { useQuery } from "@tanstack/react-query"
+import { api } from "@ui/api"
+
+export const useResolveEnsName = (name?: string) => {
+  // check if name is something we can look up
+  const lookupName =
+    // don't look up undefined
+    name !== undefined &&
+    // don't look up ethereum addresses
+    !isEthereumAddress(name) &&
+    // only look up potential ens names (dot separated string e.g. `ens.eth`)
+    isPotentialEns(name)
+      ? name
+      : undefined
+
+  // let caller detect if we're going to look name up or not
+  const isLookup = lookupName !== undefined
+
+  const { data: address, ...rest } = useQuery({
+    queryKey: ["useResolveEnsName", name],
+    queryFn: async () => {
+      if (!name || !isLookup) return null
+
+      // resolve ens name
+      return (await api.accountsOnChainIdsResolveNames([name]))[0] ?? null
+    },
+    enabled: isLookup,
+    cacheTime: Infinity,
+    initialData: () => name && ensNamesCache.get(name),
+    onSuccess: (address) => {
+      if (!name) return
+
+      // update cache
+      if (address === undefined) ensNamesCache.delete(name)
+      else ensNamesCache.set(name, address)
+
+      // persist cache to local storage
+      persistEnsNamesCache()
+    },
+  })
+
+  return [address, { isLookup, ...rest }] as const
+}
+
+const ensNamesCache = new Map<string, string | null>(
+  JSON.parse(localStorage.getItem("TalismanEnsNamesCache") ?? "[]")
```

Comment:
feeling this would fit better in a dexie table as it's "throwable data" and wouldn't block ui thread, though i understand that the simplicity of using localStorage is a benefit.
curious to know @chidg opinion on this ?

Reply:
Throwable in the sense that this could throw an error? Personally I am not a big fan of indexedDB and dexie in general, it has some specific features that are valuable for some use cases for us (subscriptions, big storage) but it generates a lot of errors that are hard to troubleshoot and handle. For small data objects which don't change very often like this, I think localstorage is a good solution. Is there another way we can get around the blocking issue @0xKheops ?

Reply:
i meant throwable data as data that can be deleted and regenerated in case there is too low storage space.
it's a similar case as balances and other data that we put in indexed db currently, where wallet doesn't mind if browser purges the db.

The ui blocking is because local storage read/writes are synchronous, not a huge deal tho as cache is loaded only once each time the dashboard or popup is opened. Might be worth checking if there's a perf hit with 20+ accounts but probably not. 
workaround would be dexie or extension storage which use asynchronous read/writes, or just no  caching at all on frontend appart from useQuery's : maybe in-memory cache on backend side would be sufficient

it was mainly food for thoughts, current implementation works.

Reply:
I think this is more likely to become a problem if the cache grows forever, so I've made a fix for that!